### PR TITLE
Comments 💬 (backend + frontend)

### LINKED ISSUE #1127: Comments! 💬 (front end)
### Description

Time to start adding comments to the platform! This is a big thing, its the first time we allow community interaction!
You can find more information in our [V2 Document ](https://docs.google.com/document/d/1hyWc2a9cn04B107UzUKd4-12T5v0VrlWmYTa8KZnYhQ/edit) (page 8) -Nested replies are skipped for now.

### Task
This issue is to get the front end in place. The buttons and the styling. The other issue #1128  is for connecting the backend.

### Page related

Will be used in :

- https://community.preciousplastic.com/how-to/page

### Mockup

Overview
afbeelding

Comments box:
afbeelding

When logged in:
afbeelding

When not logged in:
afbeelding

### LINKED ISSUE #1128: Comments! 💬 (back end)
### Description

We are introducing comments to the platform. More information on this issue #1127.
All the frontend work would be done in #1127, this issue is to connect it together in the backend.

### Task
Comments are used by how-to’s. They are chronologically sorted, latest one below. Above 10 comments there is a “load more comments” button displayed. The comment section is divided in 2 elements

### 1. New comment input field
- This element is always displayed when logged in
- If not logged in it shows a text “login to comment”
- The input field is plain text (Max 200 characters)
- The icon is based on your profile type
afbeelding

### 2. Comment field
- Comment fields are always displayed (for logged in & out users)
- Plain text from the new comment field (Max 200 characters)
- Username
- Location flag of the user (as used in [how-to overview](https://community.preciousplastic.com/how-to)
- Date of the comment
afbeelding

### Will be used in :
- community.preciousplastic.com/howto/page

### Mockup overview

afbeelding

### PR DESCRIPTION
**Pull request by tudi2d**
PR Checklist

- [x] - Latest `master` branch merged
- [x] - PR title descriptive (can be used in release notes)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Description

First implementation of the comments to be merged.

Comments can't be edited from the FE site right now - they can only be deleted. This should be ok for the first release, but could be added afterwards. Admins & comment creators are allowed to remove comments.

Combine PR #1145 & #1146

## Git Issues

Closes #1127 & Closes #1128 

## Screenshots/Videos

2KqKRqMsUG

### CODE REVIEW

**Comment by tudi2d** [comment]
@danitrod Your work made it really easy to build on! Thanks man! If you have any suggestions/things to be changed let me know :) 

@davehakkens Here is the combined version of the PRs

**Comment by davehakkens** [comment]
yeahh nice work @tudi2d and @danitrod.

Few comments from my side after some testing
Important ones:
**1. Deletion bug:**  In some way I ended up deleting your very controversial skateboard comment Philipp. I wanted to delete one of my own but yours seems to be removed as well. (was not logged in as admin)
**2. Input field character limit** Currently there is no limit. Should be set to 400 characters (original was 200 but might be to little)

Minor ones
**3. Icon change** The icon you see (currently the pink hand)  should be in sync with your profile icon (workspace, collectionpoint, machineshop, member etc.) So it's the same as the one you see in the right top corner to go to your profile
**4 Change location doesn't change flag** If I change my location it changes the flag in my profilee But it doesn't change the  flag on my comments. Ideally they are changed everywhere

**Comment by danitrod** [comment]
Thank you Dave, I'll tackle on points 1 and 4 as they're more related to back-end. Point 2 I think both of us can work on, I'll add a functionality for the back-end function to crop the comment to have length 400 if it comes with more than that, but the input field in the front-end should not allow more than 400 chars as well.

Is that ok @tudi2d?

**Comment by danitrod** [comment]
@davehakkens about 1, confirmed bug, I just accidentally deleted your test comments by deleting one of mine. I'll take a look into why that's happening.

**Comment by tudi2d** [comment]
Looks like this now.

image

**Comment by davehakkens** [comment]
> Visit the preview URL for this PR (updated for commit [5e36e07](https://github.com/ONEARMY/community-platform/commit/5e36e07160349f78a5865ab780d1b7333ba1836f)):
> 
> https://onearmy-next--pr1153-comments-combined-hca4yv9i.web.app
> 
> (expires Thu, 17 Jun 2021 17:34:18 GMT)
> 
> 🔥 via [Firebase Hosting GitHub Action](https://github.com/marketplace/actions/deploy-to-firebase-hosting) 🌎

Do you test it with that url? Doesn't seem to show any changes here..
Locally I can see them, looks good! But I can't post/test anything.

But perhaps it has to do with the Pull requests stacked on top of eachother?

**Comment by tudi2d** [comment]
I adjusted the styles - should look like intended now. Not sure about the error with posting things - @danitrod any idea?
image

**Comment by davehakkens** [comment]
nice looking good. One last thing for you @tudi2d Would be nice if you could align the text and remove the text limed when not logged in
afbeelding

**Comment by danitrod** [comment]
> I adjusted the styles - should look like intended now. Not sure about the error with posting things - @danitrod any idea?
> image

How did you get that error? I'm not being able to reproduce here

**Comment by tudi2d** [comment]
> How did you get that error? I'm not being able to reproduce here

I used the version deployed on the preview URL & tried to enter a comment.

**Comment by danitrod** [comment]
how weird - for me this isn't happening; is there any more information in the error so I can better debug? Also if you could check what was the sent payload in the request from the network devtool that might help

**Comment by tudi2d** [comment]
It looked like this: 
tnCWLSckEk

I think, I fixed it for me - @davehakkens does it work for you now as well? The issue was, that the country was set to `undefined` & firebase does not like that (as `undefined` seems like a none intended, missing value - `null` is the way to go). Difficult to debug without the information indeed...


image

**Comment by davehakkens** [comment]
I'm testing with that url, not sure if that is the best way. But posting seems to work now. 
However, the country flag still doesn't update here @danitrod? 
Comments:
afbeelding
Profile:
afbeelding

**Comment by davehakkens** [comment]
Woops and I found one more thing @danitrod! 
You are now showing the "Display name" and it should show the "Username". 
so for instance `Precious Plastic Headquarters` should be `precious-plastic`  
So it's consistent throughout the platform.

Sorry for not seeing this one sooner or making it more clear in the first place. 🙏

★ as a bonus you could also make it link to the user profile. But no need. We can also do it in an issue later

**Comment by tudi2d** [comment]
I will create an issue for the link to the profile, so it can be worked on after the PR is merged. It should be a good & easy follow up.

My suggestion would be to change the "Display name" & "Username" & than let @chrismclarke review the PR maybe, as this is already becoming a bigger PR. Not sure how hard fixing the flag issue is, but in my opinion a country change is (for the most users & workspaces) probably more of an edge case. It's great, that we already cached it & if it's not easy to fix it could also be a good follow up (as it is more of an improvement, than a bug to the comments). What do you guys think?

**Comment by davehakkens** [comment]
yeh good point, getting big indeed! Your proposal sounds good to me

**Comment by danitrod** [comment]
Nice catch @tudi2d ! Definitely difficult to debug, thank you. @davehakkens about updating information when the user updates, unfortunately the Firebase functions aren't deployed with the preview as far as I know, so we can't know if it's working yet - I left a note here https://github.com/ONEARMY/community-platform/pull/1146#issuecomment-842798487. We'll need Chris' help to check on how to test that. 

I agree we this PR is getting a bit overloaded, I will update the comments to use `username` and not `displayname` and I guess we can merge after the review.

**Comment by tudi2d** [comment]
Looks good! I liked, how we worked on this really asynchronously across time zones & still build it rather fast (for a free time project) - good communication & quality work from your site @danitrod! 🙂 Thanks for the feedback & testing @davehakkens!

**Comment by chrismclarke** [comment]
Sorry, I know I've been slow to review this, have got held up by a bunch of core build issues not working correctly. Will do my best to find time this week once the rest of the build is stable again.

File: functions/src/userUpdates/index.ts [hunk 1]
```diff
@@ -1,32 +1,50 @@
...
-import { DB_ENDPOINTS, IDBDocChange, IUserDB } from '../models'
+import { DB_ENDPOINTS, IDBDocChange, IUserDB, IHowtoDB } from '../models'
 import { backupUser } from './backupUser'
 
 /*********************************************************************
  * Side-effects to be carried out on various user updates, namely:
  * - create user revision history
  * - update map pin location on change
  * - update howTo creator flag on change
+ * - update userName and flag on any comments made by user
```
Comment (chrismclarke) [inline functions/src/userUpdates/index.ts:11]: I didn't know you could even change username! I think this leaves us with a few challenges to figure out, but good to have included in the comment update. Longer term we will probably need a refactor to keep user id independent of user name (I can't remember why we didn't do that in the first place - possibly to make serving the avatar image easier)... but in any case good to have this here for now

File: src/models/howto.models.tsx [hunk 2]
```diff
@@ -11,7 +23,10 @@ export interface IHowto extends IHowtoFormInput, IModerable {
   cover_image: IUploadedFileMeta
   files: Array<IUploadedFileMeta | File | null>
   steps: IHowtoStep[]
+  // Comments were added in V2, old howto's may not have the property
+  comments?: IComment[]
```
Comment (chrismclarke) [inline src/models/howto.models.tsx:27]: Thanks for adding the comment to explain this

File: src/components/Comment/CommentTextArea.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+`
+
+export const CommentTextArea = ({ onSubmit }) => {
+  const [comment, setComment] = useState('')
+  const [loading, setLoading] = useState(false)
+  const { stores } = useCommonStores()
+  const user = stores.userStore.activeUser
+
+  async function keyDown(e) {
+    if (e.key === 'Enter' && !e.shiftKey && comment.trim().length) {
```
Comment (chrismclarke) [inline src/components/Comment/CommentTextArea.tsx:66]: Nice idea to add this, would definitely expect pressing the enter key to submit. I'm not sure if you can achieve the same functionality using simple form elements (e.g. https://stackoverflow.com/questions/33211672/how-to-submit-a-form-using-enter-key-in-react-js/33212911), but happy to keep this in to be more explicit (and add the empty check)

File: src/stores/Howto/howto.store.tsx [hunk 4]
```diff
@@ -136,6 +138,107 @@ export class HowtoStore extends ModuleStore {
     return needsModeration(howto, toJS(this.activeUser))
   }
```
Comment (chrismclarke) [inline src/stores/Howto/howto.store.tsx:140]: I think likely we will want to reuse comments in other pages, so might be better to try and separate out into its own store or set of methods. But happy to keep here for now and can always refactor later.

**Review by chrismclarke — APPROVED** [review]
Really nice job here, thanks so much for all you've done @danitrod and @tudi2d.

The codebase looks pretty solid to me, I think there'll likely a few minor UI changes required and be good to get some more user input before going live so I'll add a `beta-tester` authwrapper to keep it only viewable by admins/beta-testers, but otherwise definitely happy to merge. 

I really appreciate that you also took some time to think about challenges with things like users changing names - for me this is a major sticking point of the current database setup making these types of modules much harder to implement (than say, graphql where you could query and merge users names, countries etc. as you retrieve the comments). But I think what you've done is as good as we can hope for short term so thanks for finding a way to work in the constraints.

**Pull request merged**